feat(trim): introduce TRIM operator (#329)
IBM/ado: examples/trim/EXAMPLE_SIMPLE.md

     representative set of points to get a baseline understanding of the space.
2. **Iterative Modeling**:
   `trim` then enters a loop: it uses the data
   it has gathered to train a preliminary model (using `AutoGluon`),
    uses that model's intelligence to decide which point to sample next,
     measures that point, and then retrains the model.
   It stops automatically when it determines that
   further sampling won't significantly
   improve the model's accuracy, saving you time and resources.

Finally, it trains one high-quality model on all the data it has collected
and saves it for you to use.

## Creating a `discoveryspace`

A `discoveryspace` describes the parameters you want to
 explore (`entitySpace`) and how to measure them (`measurementSpace`).
In this example, we'll use a custom Python function
 `calculate_pressure_ideal_gas` as our experiment.

First, create the `discoveryspace` by executing this command
 from the repository root:

```commandline
ado create space -f examples/trim/configs/space_pressure.yaml --new-sample-store
```

This will create a new space and a sample store to hold the measurement results.

> It will confirm with:

Reviewer: Let's avoid the case where people think "wait, the identifier I got is different"
```suggestion
The output will be similar to:
```

```commandline
Success! Created space with identifier: space-bfed2d-19b49a
```

## Exploring with a `trim` Operation

Next, we will run an `operation` that uses `trim` to explore the `discoveryspace`.

> The configuration for our operation is in `op_pressure.yaml`:

Reviewer: `trim` -> TRIM

```suggestion
## Exploring with a TRIM Operation

Next, we will run an `operation` that uses TRIM to explore the `discoveryspace`.
The configuration for our operation is in `op_pressure.yaml`:
```

```yaml
# op_pressure.yaml
operation:
  module:
    operationType: characterize
    operatorName: trim
  parameters:
    targetOutput: pressure
    batchSize: 1
    iterationSize: 5
    outputDirectory: trim_models
    stoppingCriterion:
      enabled: true
      meanThreshold: 0.9
      stdThreshold: 0.75
    autoGluonArgs:
      fitArgs:
        time_limit: 20
        presets: medium
    noPriorsParameters:

>       targetOutput: pressure

Reviewer: Let's include the file to avoid drift

```suggestion
{% include-markdown "./configs/op_pressure.yaml" %}
```

```

To run the operation, execute:

<!-- markdownlint-disable line-length -->
```commandline
ado create operation -f examples/trim/configs/op_pressure.yaml --use-latest space
```
<!-- markdownlint-enable line-length -->

### What to Expect in the Terminal

You will see a lot of output as `trim` does its work.
Let's break down the key stages,

> in the case of no point present in the discovery space at the beginning of the operation:

Reviewer: Let's avoid a possibly confusing "no point present in the discovery space" by just removing that part. The next anchor correctly contextualises it and uses sample store.

```suggestion
You will see a lot of output as TRIM does its work. Let's break down the key
stages:
```

#### Stage 1: No-Priors Characterization

Since we started with an empty sample store,
`trim` first sees that it doesn't have enough data.

> It will log this and begin the initial characterization phase.

Reviewer: Explain what are the effects of an empty sample store
```suggestion
Since in our example we started with an empty sample store, TRIM cannot
immediately build a model. It will log this and begin the initial
characterization phase.
```

<!-- markdownlint-disable line-length -->

```commandline
2026-01-16 14:56:57,589 WARNING   MainThread           trim.utils.space_df_connector: get_df_at_least_one_measured_value: No measured properties found in the discovery space
...
2026-01-16 14:56:57,656 WARNING   MainThread           trim.operator  : trim                : Only 0 points in the source space.
Starting with no-prior characterization operation, it will sample 20 points.
```
<!-- markdownlint-enable line-length -->

It then runs a simple sampling operation
(in this case, using Concatenated Latin Hypercube Sampling or `clhs`)
to gather the initial data points.

> You will see output for each point being measured:

Reviewer: Formatting + missing comma + slight wording change for output

```suggestion
It then runs a simple sampling operation (in this case, using Concatenated Latin
Hypercube Sampling, or `clhs`) to gather the initial data points. You will see
output similar to the following for each point being measured:
```

...
(RandomWalk pid=10736) 2026-01-16 14:57:20,723 INFO      AsyncIO Thread: default trim.trim_sampler: iterator            : [Batch under consideration: 5] Training metric: root_mean_squared_error;
(RandomWalk pid=10736) Best model: NeuralNetTorch; score_val: -8.49; holdout_score: -669.00
```
<!-- markdownlint-enable line-length -->

After a set number of iterations (defined by `iterationSize`),
it will check the stopping criterion:

<!-- markdownlint-disable line-length -->

```commandline
(RandomWalk pid=10736) 2026-01-16 14:57:48,947 INFO      AsyncIO Thread: default trim.trim_sampler: iterator            : Testing stopping criterion after measuring 14 points, mean_ratio={mean_ratio} and std_ratio={std_ratio}
(RandomWalk pid=10736) 2026-01-16 14:57:48,947 INFO      AsyncIO Thread: default trim.trim_sampler: iterator            : Stopping not triggered for i=14
```
<!-- markdownlint-enable line-length -->

#### Stage 3: Stopping and Finalizing

The iterative process continues until the model's performance stabilizes.
At this point, the stopping criterion is met,

> and `trim` proceeds to train one final model on all the data it has gathered.

Reviewer: `trim`->TRIM, formatting, verb tense change
```suggestion
The iterative process continues until the model's performance stabilizes. At
that point, the stopping criterion is met, and TRIM will train one final model
on all the data it has gathered.
```

<!-- markdownlint-disable line-length -->

```commandline
Success! Created operation with identifier operation-trim-v0.1-8b23a245 and it finished successfully.
```
<!-- markdownlint-enable line-length -->

## Looking at the `operation` output

The `trim` operator saves the final trained `AutoGluon` model
to the directory specified by `outputDirectory`
 in your operation file (here, `trim_models_finalized`).
 You can now load this `TabularPredictor` in your own code
  to make predictions on any unmeasured points in your parameter space.

You can also view the entities that were sampled during the entire operation.
 `trim` actually runs two sub-operations
 (one for characterization, one for iterative modeling).
  You can see the relationship with:

```commandline
ado show related operation --use-latest
```

This will show the `discoveryspace` and the sub-operations that were run.
To see the actual data points from the final iterative phase, you can run:

```commandline
# The identifier will be different for your run
ado show entities operation randomwalk-1.3.3.dev14+177ead3c.dirty-14a313
```

This will display a table of the entities sampled and their
 measured pressure values.

> 

Reviewer: We need to be careful here: while it's true that I see another operation and the space if i do show related of the latest operation, if i do the same for the space (which is what I would've assumed I had to do), I see 3 operations: two randomwalks, and a trim operation. At that point I wouldn't know which of the two random walks I would need to check.

Should we do show entities space to avoid issues?

<!-- markdownlint-disable line-length -->
```text
             Measurements - randomwalk-1.3.3.dev14+177ead3c.dirty-14a313              
┏━━━━━━━┳━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━┓
┃ index ┃ mol  ┃ temperature ┃ volume ┃ pressure ┃ request_id ┃ entity_index ┃ valid ┃
┡━━━━━━━╇━━━━━━╇━━━━━━━━━━━━━╇━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━┩
│ 21    │ 0.40 │ 296.00      │ 8.00   │ 123.05   │ de3011     │ 0            │ True  │
│ 20    │ 0.20 │ 276.00      │ 1.00   │ 458.96   │ eec630     │ 0            │ True  │
│ 19    │ 0.20 │ 286.00      │ 2.00   │ 237.79   │ 15f0f5     │ 0            │ True  │
│ 18    │ 0.10 │ 288.00      │ 6.00   │ 39.91    │ 938036     │ 0            │ True  │
│ 17    │ 0.70 │ 278.00      │ 2.00   │ 809.00   │ b6ede7     │ 0            │ True  │

> ...

Author: solved
